influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +115,231 @@ struct CacheValue {
     Success(Arc<CacheValue>),
-    /// The request to the object store failed
-    Failed,
-    /// The cache entry was deleted
-    Deleted,
-    /// The object is too large for the cache
-    TooLarge,
 }
 
-impl CacheEntry {
-    /// Get the size of thje cache entry in bytes
+impl CacheEntryState {
+    /// Get the approximate size of the cache entry in bytes
     fn size(&self) -> usize {
         match self {
-            CacheEntry::Fetching => 0,
-            CacheEntry::Success(v) => v.size(),
-            CacheEntry::Failed => 0,
-            CacheEntry::Deleted => 0,
-            CacheEntry::TooLarge => 0,
+            CacheEntryState::Fetching(tx) => std::mem::size_of_val(tx),
+            CacheEntryState::Success(v) => v.size(),
         }
     }
+}
+
+/// A cache for storing objects from object storage by their [`Path`]
+///
+/// This acts as a Least-Frequently-Used (LFU) cache that allows for concurrent reads. See the
+/// [`Cache::prune`] method for implementation of how the cache entries are pruned. Pruning must
+/// be invoked externally, e.g., on an interval.
+#[derive(Debug)]
+struct Cache {
+    /// The maximum amount of memory this cache should occupy
+    capacity: usize,
+    /// The current amount of memory being used by the cache
+    used: AtomicUsize,
+    /// The maximum amount of memory to free during a prune operation
+    max_free_amount: usize,
+    /// The map storing cache entries
+    ///
+    /// This uses [`IndexMap`] to preserve insertion order, such that, when iterating over the map
+    /// to prune entries, iteration occurs in the order that entries were inserted. This will have
+    /// older entries removed before newer entries
+    map: RwLock<IndexMap<Path, CacheEntry>>,
+}
 
-    fn is_fetching(&self) -> bool {
-        matches!(self, CacheEntry::Fetching)
+impl Cache {
+    /// Create a new cache with a given capacity and max free percentage
+    ///
+    /// The cache's `max_free_amount` will be calculated using the provided values
+    fn new(capacity: usize, max_free_percentage: f64) -> Self {
+        Self {
+            capacity,
+            used: AtomicUsize::new(0),
+            max_free_amount: (capacity as f64 * max_free_percentage).floor() as usize,
+            map: RwLock::new(IndexMap::new()),
+        }
     }
 
-    fn is_success(&self) -> bool {
-        matches!(self, CacheEntry::Success(_))
+    /// Get an entry in the cache or `None` if there is not an entry
+    ///
+    /// If the entry is in `Fetching` state, then this will await the etry having been fetched.
+    async fn get(&self, path: &Path) -> Option<Arc<CacheValue>> {
+        let map = self.map.read().await;
+        let entry = map.get(path)?;
+        match &entry.state {
+            CacheEntryState::Fetching(tx) => {
+                let mut rx = tx.subscribe();
+                // TODO(trevor): is it possible that the sender has sent the result at this point?
```

Comment:
yeah, I'm not sure if this will do what you want it to do. A Receiver would be better and you can wait until there's `Some` in there. Basically, what I was thinking is you wait on the changed on the receiver and when that's cleared, you then grab a read lock on the cache again and get the value.

Or maybe it would work to bring down the actual value.

Another option would be for fetching to be `CacheEntryState::Fetching(Arc<RWLock<Option<Arc<CacheValue>>>>)`. Then, when you create that lock, grab a write lock right away, and then set the cache state. Then when the thing returns from object store, set the CacheValue in the cache map, and also set that locked value to the arc of the cache value, then releasing the write lock.

For any cache accessors that hit it while fetching, they just try to acquire a read lock on the fetching lock and await it. Once they get the read lock, it's because the cache value is in there and they can retrieve it.

Reply:
I took a slightly different approach by having the `Fetching` state store a shared future directly, so if it gets requested while fetching, it can just await the future. This was in https://github.com/influxdata/influxdb/pull/25389/commits/667eab588b80adf8049a5bec8cf43a0612d47560.